models/pull.go

```diff
@@ -176,6 +177,39 @@ func (pr *PullRequest) loadProtectedBranch(e Engine) (err error) {
 	return
 }
 
+func (pr *PullRequest) requiredApprovals() int64 {
+	if pr.ProtectedBranch == nil {
+		if err := pr.loadProtectedBranch(x); err != nil {
+			log.Error("Error loading ProtectedBranch", err)
+		}
+	}
+	if pr.ProtectedBranch != nil {
+		return pr.ProtectedBranch.RequiredApprovals
+	}
+	return 0
+}
+
+// RequiresApproval returns true if this pull request requires approval, otherwise returns false
+func (pr *PullRequest) RequiresApproval() bool {
+	return pr.requiredApprovals() > 0
+}
+
+// GetReviewLabel returns the formatted review label with counter for the review of this pull request
+func (pr *PullRequest) GetReviewLabel(lang string) string {
+	if pr.RequiresApproval() {
+		rejections := pr.ProtectedBranch.GetRejectedReviewsCount(pr)
+		if pr.ProtectedBranch.GetRejectedReviewsCount(pr) > 0 {
+			return i18n.Tr(lang, "repo.pulls.review_rejected", rejections)
+		}
+		approvals := pr.ProtectedBranch.GetGrantedApprovalsCount(pr)
```

Comment:
I think that a PR that has been repeatedly approved and rejected from one reviewer should count only one approval or one rejection, depending on which happened last:

* **Reviewer 1**: approves, requests changes = **0 approvals**, **1 change requests**
* **Reviewer 2**: approves = **1 approval**, **0 change requests**
* **Reviewer 3**: requests changes, requests changes, requests changes, approves = **1 approvals**, **0 change requests**
* **Reviewer 4**: requests changes = **0 approvals**, **1 change requests**

**Grand total: 2 approvals, 2 change requests**
 
To achieve that, IMHO all counts (approvals, rejections) should be processed in the same database query. Check [this comment](https://github.com/go-gitea/gitea/pull/9055#discussion_r350977864) (and other reviews of mine in that PR) for examples on how to achieve that.

Reply:
I think that is already achieved. Both `GetGrantedApprovalsCount` and `GetRejectedReviewsCount` only counts reviews that are _official_, and official reviews are only the latest review (approval or rejection) of users that are in the whitelist.

Reply:
Duh, I forgot about the `official` property! I had a rough week... I need some rest. 😄

Reply:
We could use different color tag for approved counter and requested counter. Even the approval requirement. `1/2` means this PR require 2 approvals but has 1 approved.

Reply:
One should approve or request change.